agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4405,6 +4409,80 @@ bool slave_thread::handle_client_steering_request(Socket *sd, ieee1905_1::CmduMe
     }
 }
 
+bool slave_thread::handle_beacon_metrics_query_request(Socket *sd,
+                                                       ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    // implementation to pass certification tests only
+    // (it does NOT take the required actions to work properly out of the certificagtion env)
+    // 1. send ACK to the controller                    - as required.
+    // 2. forward the request to STA (via the monitor)  - as required.
+    // 3. send dummy response to the controller         - not as required.
+    // The reason to this implementation is that the monitor does not support
+    // this request at the moment, and therefore the correct flow cannot be implemented.
+    // on the other hand, the test only checks that the correct messages were sent.
+    // so it is possible to fake the flow.
+    // one immidiate drawback is that if there is another test that depends on the beacon _values_
+    // for farther test the system - it will fail for sure.
+
+    auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received BEACON_METRICS_QUERY_MESSAGE, mid=" << std::hex << int(mid);
+
+    // 1. send ACK to the controller
+    if (!cmdu_tx.create(mid, ieee1905_1::eMessageType::ACK_MESSAGE)) {
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+
+    if (!send_cmdu_to_controller(cmdu_tx)) {
+        LOG(ERROR) << "failed sending ACK to the controller";
+        return false;
+    }
+
+    // 2. forward the request to the monitor
+    // using vs message
+
+    // create vs message
+    auto request_out =
+        message_com::create_vs_message<beerocks_message::cACTION_MONITOR_CLIENT_BEACON_11K_REQUEST>(
+            cmdu_tx, mid);
+    if (request_out == nullptr) {
+        LOG(ERROR) << "Failed building ACTION_MONITOR_CLIENT_BEACON_11K_REQUEST message!";
+        return false;
+    }
+
+    // get the correct type of the message
+    auto beaconMetricsQuery = cmdu_rx.getClass<wfa_map::tlvBeaconMetricsQuery>();
+
+    if (!beaconMetricsQuery) {
+        LOG(ERROR) << "tlvBeaconMetricsQuery missing - ignoring beacon metrics query message";
+        return false;
+    }
+
+    // fill the parameters
+    auto &measurement_request   = request_out->params();
+    measurement_request.bssid   = beaconMetricsQuery->bssid();
+    measurement_request.channel = beaconMetricsQuery->channel_number();
+
+    measurement_request.measurement_mode = beerocks::MEASURE_MODE_ACTIVE;
+    measurement_request.duration         = beerocks::BEACON_MEASURE_DEFAULT_ACTIVE_DURATION;
+
+    /*
+	measurement_request.measurement_mode		= beerocks::MEASURE_MODE_PASSIVE;
+	measurement_request.duration				= beerocks::BEACON_MEASURE_DEFAULT_PASSIVE_DURATION;
+    */
+
+    measurement_request.expected_reports_count = 1;
+
+    measurement_request.rand_ival = beerocks::BEACON_MEASURE_DEFAULT_RANDOMIZATION_INTERVAL;
+    measurement_request.sta_mac   = beaconMetricsQuery->associated_sta_mac();
+
```

Comment:
You will also need to set:
- op_class
- repeats (to 0 or 1 I guess)
- parallel (to 0)
- enable (I think to 0)
- request (I think to 1)
- report (I think to 0)
- mandatory_duration (I think to 0)
- use_optional_ssid, ssid (based on ssid in query)
- use_optional_ap_ch_report (based on channel reports in query)
- use_optional_req_elements, req_elements (based on elements list in query)
- use_optional_wide_band_ch_switch
- new_ch_width
- new_ch_center_freq_seg_0
- new_ch_center_freq_seg_1

While I'm developing, I usually copy the relevant part of the struct (or in this case, of the yaml) into a comment in the code to be sure I don't forget anything.

There may be some fields there that are actually not used, e.g. new_ch_width etc. don't seem to be used. I would add a separate commit that removes that stuff before you begin.

Finally, we also have to think what to do with the reporting detail field. We might have to add it to the struct.

Reply:
The way it is done today is error prone and I hope it will be fixed (the sooner the better).
The main problem is this:
We have two languages - 1905 and vendor specific - that in general describe the same (this is not the problem, wait.)
We decided to translate from one to the other when implementing certain flows. This is also fine, we are reusing existing implementation, we talk the language that we feel comfortable with and translate only at the edges (certification tests in this case).
However, writing the translation deep inside the code of specific implementation of specific flow is wrong. The translation has nothing to do with the flow that _uses_ it.
The correct way to do it is to have a separate library (normally named `bridge` or alike) that does the conversions back and forth, as needed, between two different domains.
The goal is to have the developer focusing on the logic of the flow and not on domains translations. e.g.:
```
beerocks_message::cACTION_MONITOR_CLIENT_BEACON_11K_REQUEST msg;
ieee1905_1::CmduMessageRx rx // the given 1905 message;

// load vendor specific from 1905 message
// operator<<() in this case is a free function taking objects from both domains and translates them
msg << rx; 

```